### ISSUE TITLE
Enhancement: [no-unused-vars] Add option to allow variables to be used only as types or allow 'declare'

### ISSUE THREAD

**Issue by InExtremaRes** [typescript-eslint/typescript-eslint#10266]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.6.2&fileType=.ts&code=CYUwxgNghgTiAEYD2A7AzgF3gfWwIxihWAC54BXFASwEdyE0BPAWzyQgG4AoEADwAckMLBkb8EAIULEQwAMoYYVFAHN4AXniYlq%2BADJ4Abxz5ppeKPFIAZiYJFg8AL7cgA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tidmUQAmtAG4BDaKgyRE0aB2iRwYAL4hlQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A&tokens=false

### Repro Code

```TypeScript
declare const __brand: unique symbol;
type BrandedString = string & { __brand: typeof __brand };
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unused-vars": "error",
  },
};
```

### Expected Result

Have an option to either allow variables declared using `declare` or to allow variables as types in general.

### Actual Result

Variable `__brand` is flagged as unused without much options (besides disabling the rule or use a pattern).

### Additional Info

In #9330 the rule now detects and flags variables that are used only as types. I propose to add an option to disable this behavior, which can be enabled by default.

I've seen some discussion about cases where it's desirable to support such a thing in #9275 and #10184; the latter was closed since the particular requirement in that issue can be achieved by other means. A particular case where I encounter the issue was trying to *brand* a type with an unique type. I would like to do something similar to this:

```ts
type BrandedString = string & { __brand: unique symbol };
```

However that's not possible since an `unique symbol` cannot be used that way, but I can do this:

```ts
declare const __brand: unique symbol;
type BrandedString = string & { __brand: typeof __brand };
```

This way the `BrandedString` will be different to any other `BrandedString`, even if it is declared the same way since every symbol is actually unique. However this case is flagged by the rule.

(In the actual code I use a declared private field so the `#__brand` property and its type cannot be accessed from anyone else, but my point still remains)

I propose to either allow cases where the variable is declared using `declare` or adding a new option to allow variables to be used in type positions. The first case has been touched in the mentioned issues and will satisfy this particular use case, but I think adding an option is the simplest and more universal solution.

**Comment by Josh-Cena** [comment]
Agreed, I think we should find some way to make this use case work.

**Comment by bradzacher** [comment]
My 2c - this usecase isn't broadly common enough to warrant the need for an entire option to handle it, nor is it worth us expanding bandwidth/maintenance burden on it.

Across the entirety of public github repos there are **only** [8.3k hits](https://sourcegraph.com/search?q=context:global+/declare%5Cs%2B%28const%7Clet%7Cvar%29%5Cs%2B.%2B%3F%5Cs*:%5Cs*unique%5Cs%2Bsymbol/+language:TypeScript+count:all&patternType=keyword&sm=0).

IMO document, disable comment and move on is all we should fo..

**Comment by Josh-Cena** [comment]
> My 2c - this usecase isn't broadly common enough to warrant the need for an entire option to handle it, nor is it worth us expanding bandwidth/maintenance burden on it.

It's a totally valid use case though, most people just never found the need to use branded types. You would probably find similarly much usage of `Reflect`—doesn't mean these users are simply ignored.

In general, `declare` itself is a type-only construct, and more often than not actually cannot be used at runtime.

**Comment by bradzacher** [comment]
I'm not saying ignore. Documenting the usecase as something we know exists but won't add logic for is not ignoring them.

We have limited bandwidth for building, reviewing and maintaining -- and we need to be realistic about that limitation. In an ideal world we would handle all cases but in reality we physically cannot.

8.3k instances is truly minuscule in the grand scheme of things -- hence I say it's definitely not worth thinking too much about.

**Comment by jakubmazanec** [comment]
Another data point: ability to turn off error "XYX is assigned a value but only used as a type" is definitely something that I would use.

**Comment by RebeccaStevens** [comment]
I'd like `no-unused-vars` to not report type imports that are referenced in JSDoc.
TypeScript itself, see the type as used, so `no-unused-vars` should probably work the same way.

Here's a small example:

```ts
import type { Foo } from "./foo";
//            ~~~ 'Foo' is defined but never used.

/**
 * JSDoc with reference to {@link Foo}.
 */
export function bar() {
  // ...
}
```

Tools like [typedoc](https://typedoc.org/) need that import to link the reference correctly.

**Comment by bradzacher** [comment]
@RebeccaStevens that is off topic for this thread and also is a duplicate of https://github.com/typescript-eslint/typescript-eslint/issues/9435 and all the other threads listed there. We do not support JSDoc in this plugin.

**Comment by falkenhawk** [comment]
imo oxlint got it right: https://oxc.rs/docs/guide/usage/linter/rules/eslint/no-unused-vars
> A variable `foo` is considered to be used if any of the following are true:
> (...)
> It is used as an operand to TypeScript's `typeof` operator (`const bar: typeof foo = 4`)

A `typeof` over a variable should mark it as "used", regardless of what's the resulting, compiled js code is. Or please add a config option for it 🙏

**Comment by JoshuaKGoldberg** [comment]
> imo oxlint got it right

Ok, great, go use oxlint 😛

**Comment by JoshuaKGoldberg** [comment]
Answering more seriously: what I'm personally looking for here is _good_ motivation _why_ you'd want this option.

So far the use cases that I know of are:

* `declare const` and branding symbols: as in the OP
* Reflect: https://github.com/typescript-eslint/typescript-eslint/issues/10266#issuecomment-2453259389
* `typeof` operators on complex values

---

For `declare const` and branding symbols: this is a pretty uncommon use case, and something that's working outside the bounds of the type system. I've only exceedingly rarely actually seem someone do that strategy in real-world code - and I did a _lot_ of investigation work in writing https://www.learningtypescript.com/articles/branded-types. The reason I opted not to mention symbols in that article is that they're _really, really rare_.

I also don't think you need a standalone `declare`d variable in this use case? You can inline the type as just `unique symbol`. [This playground shows two brands being used as `readonly __brand: uniqye symbol`](https://typescript-eslint.io/play/#ts=5.6.2&fileType=.ts&code=FAUwHgDg9gTgLgAjgTwiBAhGBDAdgExHwDkBXAWwCMQYEBeBAZzhgEtcBzBAMgQG8EMENnxRcAG2QIA%2BtMo4CALgSlcrAI6l0jZFSjiEAXwDcwUJFiIUaTAsL4Ayi3ZcGzNpx79Bw0RKmy8nj4yqoaWky6lPpGpsCEAMbi2EII4iCIAG7Y4lpkVDTKWMFE%2BdQwcYnJqelZOVpOHhxFdkSNLnHZuSBlNPQIXQ3OnKZAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tidmUQAmtAG4BDaKgyRE0aB2iRwYAL4hlQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A&tokens=false).

---

For Reflect, does anybody have a use case they can share? If so, does it actually need a `declare` or `typeof`, or can it be inlined the way the previous playground can be?

---

For `typeof` operators on complex values, I again have yet to see a use case where this is actually useful. The most common request I've seen around it has been for Zod schemas or similar:

```ts
import { z } from "zod";

const mySchema = z.object({ value: z.number() });
//    ~~~~~~~~
// Unused runtime variable (only used in the typeof)

export type Shape = z.infer<typeof mySchema>;
```

...but that variable _is_ unused! The `type Shape` is just a type, which you can declare without the extra work of going through Zod:

```ts
export interface Shape {
  value: number;
}
```

---

All that is to say: we get that this area is something that has feature requests often. See:  #9679, #9697, #9778, #9811, #10184 (n.b. not all are exact duplicates). But _most_ of the requests to allow type-only descriptions of variables have not actually been the right way to solve the use cases as presented. Unless someone can post a reasonable reason why to do this, we'd have to lean towards treating this as a _documentation_ issue rather than _feature_.

**Comment by falkenhawk** [comment]
Thanks a lot for the write-up!
...fwiw, here goes my case, which could probably be refactored but I have no idea or time for it at the moment, so I'd rather just adjust the lint config to allow the typeof to mark the var as unused and move on, instead of slapping on yet another eslint-ignore 🙈
```typescript
export type Library<TItemNames extends readonly string[]> = {
  getId: (name: TItemNames[number]) => number;
};

const allDeckNames = <const>['deck1', 'deck2'];
const allTopicNames = <const>['topic'];
const allLibItemNames = <const>[...allTopicNames, ...allDeckNames]; // "used" only in `typeof`, violates @typescript-eslint/no-unused-vars

const lib: Library<typeof allLibItemNames> = /* lib gets populated with a proxy to access items by their names */
const id = lib.getId('deck1');
const otherId = lib.getId('topic');

/// meanwhile, some logic follows where allDeckNames, allTopicNames are used
// `allLibItemNames` remains unused (only as a typeof of a merged arrays, created only for convenience)
const tests = setupGroupsTest({ decks: allDeckNames, topics: allTopicNames ...
```

You are right, in such cases the var oin fact remains unused in the compiled JS code, but we are in the context of typescript - and in this context the var is, well, used, kind of 😅  And we care about the DX, right? and not CX 😅

**Comment by JoshuaKGoldberg** [comment]
> which could probably be refactored

Indeed it can:

```ts
export type Library<TItemNames extends string> = {
  getId: (name: TItemNames) => number;
};

type AllDeckNames = 'deck1' | 'deck2';
type AllTopicNames = 'topic';
type AllLibItemNames = AllDeckNames | AllTopicNames;

declare const lib: Library<AllLibItemNames>;
const id = lib.getId('deck1');
const otherId = lib.getId('topic');
```

> but I have no idea or time for it at the moment

Please pardon my bluntness, but that to me emphasizes that this is a _"you're holding it wrong"_ (and the docs could help you adjust) situation.

The original version of the code declares three variables unnecessarily and uses two array spreads in one of them. A new reader to the code could easily get confused as to where those variables are used. "Surely if they're declared as runtime values declared, they're used somewhere?"

> we care about the DX, right? and not CX

Indeed we do. DX is a wide umbrella and there are a lot of ways to optimize for it. One of the many points we care about is making sure the tools are maximally helpful in steering code authors away from "bad patterns" in code (however that gets defined).

**Comment by falkenhawk** [comment]
Thanks for the free consultancy on coding, I appreciate it, and apologies for taking up your time with this.
So the `AllDeckNames` is now a `type`. But the whole point of (stripped down) example was that the `allDeckNames` were a readonly array of values, available at runtime 😅 The whole thing was based on that typescript can infer types from values, but it's difficult or impossible to do it the other way around, without code repetition or losing some of the dynamic features. Or am I missing something --

Surely you may guess the case - it's not my code, it's as I found it while trying to update eslint to v9, my job here is just that and fix small issues along the way, and not rewriting the logic from ground up. There is not only this one occurrence but 10 other files with code based on similar assumptions, so it's either going to be `eslint-ignore` galore or disable the `no-unused-vars` rule completely, (which I don't want to do) or postpone the eslint v9 update until I or more possibly someone else presents better, less convoluted and more spot-on use cases 😅

**Comment by JoshuaKGoldberg** [comment]
> But the whole point of the (stripped down) example was that the allDeckNames were a readonly array of values, available at runtime 😅 

In that case you can always mix-and-match:

```ts
// (these two are used elsewhere)
const allDeckNames = <const>['deck1', 'deck2'];
const allTopicNames = <const>['topic'];

type AllDeckNames = typeof allDeckNames[number];
type AllTopicNames = typeof allTopicNames[number];
type AllLibItemNames = AllDeckNames | AllTopicNames;
```

Besides the joy of refactoring someone else's code 😄, I think part of the use in going through your example is that it's a very good example of the common request behind this issue.

**Comment by jakubmazanec** [comment]
I have a use-case for the "option 3", i.e. a variable should considered used if it's used via `typeof` operator: testing types. Here is an example:

```
test('the return type is correct', () => {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars -- false positive
  let result = parseArguments([], {
    commands: ['do stuff', 'undo stuff'],
    parameters: [
          {
            description: 'String parameter.',
            type: 'string',
            choices: ['foo', 'bar'],
          },
        ],
  });
  let isTypeCorrect: IsEqual<
    typeof result, // `result` variable IS actually used
    {
      readonly command: 'do stuff' | 'undo stuff' | undefined;
      readonly parameters: readonly ['bar' | 'foo' | undefined];
      readonly options: null;
      readonly unknownOptions: null;
      readonly rest: string[];
      readonly errors: Error[];
    }
  > = true; // if the types don't match, TypeScript complains about `true` being assigned to variable typed `false`

  expect(isTypeCorrect).toBeTruthy(); // this test is only for typechecking
});
 ```
 
I'm testing if the the return type of `parseArguments` has a specific structure. I don't care how is the test actually compiled, it isn't in the resulting package.

**Comment by kirkwaiblinger** [comment]
I guess type-testing APIs get around this neatly by taking the variable whose type is to be tested as a(n) (unused) runtime value: 

https://stackblitz.com/edit/vitest-dev-vitest-73gsuqlf?file=test%2Ftype.test-d.ts

```ts
import { expectTypeOf, test } from 'vitest'

declare function parseArguments(...x: any): unknown;

test('the return type is correct', () => {
  let result = parseArguments([], {
    commands: ['do stuff', 'undo stuff'],
    parameters: [
          {
            description: 'String parameter.',
            type: 'string',
            choices: ['foo', 'bar'],
          },
        ],
  });

  expectTypeOf(result).toEqualTypeOf<{
    readonly command: 'do stuff' | 'undo stuff' | undefined;
    readonly parameters: readonly ['bar' | 'foo' | undefined];
    readonly options: null;
    readonly unknownOptions: null;
    readonly rest: string[];
    readonly errors: Error[];
  }>();
});
```

https://github.com/mmkal/expect-type

**Comment by bradzacher** [comment]
I would agree with Josh here. So far we've seen one entirely valid usecase that's very niche, one usecase that should be refactored to eliminate the unused variables, and one usecase that existing solutions.

It looks like there's a majority maintenance team in favour of just improving the documentation to show the better patterns (for the latter cases), and accepting the first case as a niche edge case that will remain unhandled, but documented as unhandled.

Accepting PRs to improve the docs!

**Comment by jakubmazanec** [comment]
That doesn't mean my use case isn't valid - `typeof` usage is still a usage and there should be a flag for it. If the argument against implementing such flag is about "useless variables after compiling", tests with `expectTypeOf` still satisfy such argument.

**Comment by bradzacher** [comment]
We never said it wasn't valid. We said there were other, established patterns which are _generally_ accepted as better and work with the rule. Our recommendation is to use these patterns instead and are what we shall support in the rule.

It's worth noting that this project is maintained by volunteers. We do not seek to cover and support 100% of patterns and there will always be cases that we intentionally keep unhandled if we do not deep such cases to be broadly applicable enough.

**Comment by cakoose** [comment]
My use case is similar to the Zod use case given in https://github.com/typescript-eslint/typescript-eslint/issues/9697 and summarized again in a comment on this issue: https://github.com/typescript-eslint/typescript-eslint/issues/10266#issuecomment-2551361581. It's slightly different in that my schemas aren't auto-generated.

My use case:
1. I have some modules that use Zod schemas as the primary way to define certain interfaces.
2. Some of those Zod schemas end up being used for runtime validation, some are just for the types.
3. For consistency, it's nice that all the interface types are specified as Zod schemas, rather than having some of them in Zod and some of them in plain TS.

I can add "// eslint-disable-next-line" suppressions or change the identifiers to match the ignore pattern.
- Downside: If the value later become come completely unused, ESLint will not report an error.

Some ideas for supporting this use case better:
1. Enhance the rule to allow me to perform more precise suppression:
    - A more precise comment `// eslint-disable-next-line @typescript-eslint/value-only-used-as-type`
    - In the rule config, define another ignore pattern, specifically for values that are only used as types.
2. Add rule config to let me ignore values-only-used-as-types that have a specific static type, e.g. `ZodType`.
3. Add rule config to let me ignore _all_ values-only-used-as-types. (Not ideal, but for our codebase we'd take this over nothing.)

Do any of those ideas sound reasonable? I'm happy to start working on a PR if the maintainers are generally on-board with one of the approaches.

**Comment by kirkwaiblinger** [comment]
> My use case is similar to the Zod use case given in #9697 and summarized again in a comment on this issue: [#10266 (comment)](https://github.com/typescript-eslint/typescript-eslint/issues/10266#issuecomment-2551361581). It's slightly different in that my schemas aren't auto-generated.
> 
> My use case:
> 
> 1. I have some modules that use Zod schemas as the primary way to define certain interfaces.
> 2. Some of those Zod schemas end up being used for runtime validation, some are just for the types.
> 3. For consistency, it's nice that all the interface types are specified as Zod schemas, rather than having some of them in Zod and some of them in plain TS.

One thing I'm curious about with this pattern is that I'm guessing you're just using the zod schema _within_ a module? 

I can see why something like this would help with consistency (which isn't reported by the rule):
```ts
// schemas.ts
import { z } from 'zod';

export const schema1 = z.object(/* etc */);
export type Schema1 = z.infer<typeof schema1>;

export const schema2 = z.object(/* etc */);
export type Schema2 = z.infer<typeof schema2>;

// some-file.ts
// schema1 is used for its value, Schema2 for its type.
import { schema1, type Schema2 } from './schemas.ts';

// also can do this if you want to do the `z.infer<typeof schema2>` or similar in this file.
import { schema1, type schema2 } from './schemas.ts';
```

but I have a harder time following why you'd want to define and use a type using a zod schema _within_ a module, where it simply either is or isn't used. Curious for more info.

**Comment by kirkwaiblinger** [comment]
(see also https://github.com/typescript-eslint/typescript-eslint/issues/9679)

**Comment by cakoose** [comment]
Yes, it's within a module.

To try and roughly describe the situation, each file is a logical "integration" with some other service. It has some endpoints it makes HTTP requests to and some endpoints it receives HTTP requests from. We use Zod to define all the types involved in those endpoints.

We use Zod at runtime for incoming data:
- Parsing an incoming HTTP request body.
- Parsing the HTTP response body for a request we made.

But the Zod schemas that are only involved in outgoing data, we don't currently use the runtime validation features. We just use the static type.

**Comment by bradzacher** [comment]
> For consistency, it's nice that all the interface types are specified as Zod schemas, rather than having some of them in Zod and some of them in plain TS. 

I'm not quite sure I understand why you'd use zod to define a type and then not use zod at runtime. That seems like needless complexity and indirection for the sake of a slight bit of consistency.
Especially when the types are module local and not exported.

> Enhance the rule to allow me to perform more precise suppression

This is not physically possible. Disable comments are by rule ID only. So to do this the check would need to be moved to a new rule.

**Comment by cakoose** [comment]
> I'm not quite sure I understand why you'd use zod to define a type and then not use zod at runtime. That seems like needless complexity and indirection for the sake of a slight bit of consistency.

- There are three kinds of definitions: (1) ones used only in API inputs, (2) ones used only in API outputs, (3) and ones used in both.
- We would like to use Zod to runtime validate API inputs and outputs, but because of some Zod peculiarities, it only really works well for inputs.
- This means we use Zod for 1 and 3, but not 2.
- However, it feels natural to write all three in Zod, rather than write 1 and 3 in Zod and write 2 in plain TypeScript.
- Things sometimes switch from category 2 to 3 or vice versa. When they do, it's nice that they're all in Zod so we don't have to rewrite the type definitions (or dig them out of the Git history).

> > Enhance the rule to allow me to perform more precise suppression
>
> This is not physically possible. Disable comments are by rule ID only. So to do this the check would need to be moved to a new rule.

Yes, the first sub-bullet (suppression comment) requires splitting up this rule.
- Precedent: [eslint-plugin-imports](https://github.com/sweepline/eslint-plugin-unused-imports) does this with the `@typescript-eslint/no-unused-vars` rule. In that case it's to allow more precise autofix rather than more precise suppression.

Another option would be if ESLint allowed rules to specify custom suppression categories.
- Seems like a long-shot, but I wonder if there have been other requests for this feature. Especially since it sort of makes sense that it sometimes makes sense for two categories of errors to be detected by the same piece of code.

The second sub-option (suppression regex) would not require splitting the rule.

**Comment by kirkwaiblinger** [comment]
I hate to say it, since I know it's really unsatisfying to hear, but I'm not personally swayed by this, at least not on its own.

- You're asking the linter to behave based on how your codebase may be in the future, rather than how it is now. That's pretty against the grain for no-unused-vars (after all - any variable might end up being required in the future). In particular, this is not even about how your codebase as a whole may be in the future, but how your _specific module-local behavior_ may look in the future. If this were about an unused export, the rule wouldn't be reporting, of course.
- You're using a zod schema to define a type, but not using it at runtime, _because you say it has problems actually being used at runtime_. If that's the case, then I'd think it's a hazard to leave that schema around, available to be used at runtime by an unsuspecting developer. Wouldn't it be safer for it to look like 
   ```ts
   /** @deprecated */
   // eslint-disable @typescript-eslint/no-unused-vars -- we're using this to define a type only but it's not safe to be used at runtime!!
   const outGoingRequestSchemaDoNotUse = z.object(/* ... */);
   type OutGoingType = z.infer<typeof outGoingRequestSchema>; // or maybe z.input<typeof outGoingRequestSchema>
   ```
   or even better, confining the schema to its own module that only exports a type (_and therefore can get tree shaken, so the schema creation side effect will never be executed!_)
   ```ts
   // models.ts
   
   // <disable comment with explanation>
   const outGoingRequestSchemaDoNotUse = z.object(/* ... */);
   export type OutGoingType = z.infer<typeof outGoingRequestSchema>; // don't forget z.input<> could be different!
   ```
- zod schemas aren't 1:1 with TS types. There is no way to accurately reflect a `z.object({ }).strict()` for example. So you may have a doubly difficult time ever using the runtime schemas that you've provided since your valid TS isn't necessarily a valid way to fulfill the schema. 
- The git diff to switch between a straightforward zod schema and a TS type is pretty trivial in the cases that I've done it.

So, I'm still on the side that a conspicuous suppression comment would be a bigger benefit to your codebase than to permit this pattern unchecked 🫤 

***

I personally think an appropriate order of operations is 
1. Update documentation with known patterns where users may encounter the used-only-as-a-type report and what potentially preferable alternative strategies exist in those cases.
2. If we get significant user feedback thereafter we can do our "evaluating community engagement" process

**Comment by cakoose** [comment]
> I hate to say it, since I know it's really unsatisfying to hear, but I'm not personally swayed by this, at least not on its own.

Totally fine! Appreciate you taking the time to read and respond.

> You're asking the linter to behave based on how your codebase may be in the future, rather than how it is now.

Yes, but I don't think that's without precedent? I think a sort of similar thing happens with imports.
- Many codebases allow using a plain `import {SomeClass} ...` even if `import type {SomeClass} ...` is all you need. 
- I've worked in codebases that allow importing a commonly-used utility library (e.g. lodash) even if it's not currently used in the module.

I think part of the reason for those is to avoid import churn when making minor changes to the module's code.

In my case the churn is slightly more annoying -- having to switch a between Zod and TS type definition. It's not just about the work, but I can imagine a bug being introduced in the process.

If I could suppress more precisely, the churn would just be adding/removing the suppression comment.

> You're using a zod schema to define a type, but not using it at runtime, _because you say it has problems actually being used at runtime_. If that's the case, then I'd think it's a hazard to leave that schema around, available to be used at runtime by an unsuspecting developer. Wouldn't it be safer for it to look like 

There's nothing about _that particular schema_ that is unsafe. It's just generally unsafe to use any Zod schema for output validation. We only use them for input validation.

> or even better, confining the schema to its own module that only exports a type (_and therefore can get tree shaken, so the schema creation side effect will never be executed!_)

Yup, that has some advantages. But the downside is that, since the `type` is exported, we will no longer get a lint error when that whole thing (type and Zod schema) become unused.

> * zod schemas aren't 1:1 with TS types. There is no way to accurately reflect a `z.object({ }).strict()` for example. So you may have a doubly difficult time ever using the runtime schemas that you've provided since your valid TS isn't necessarily a valid way to fulfill the schema.

Is that a problem specific to this use case? Or an argument against using Zod schemas / `z.infer` in general?

> So, I'm still on the side that a conspicuous suppression comment would be a bigger benefit to your codebase than to permit this pattern unchecked 🫤

I'm on board with that option, except with the ability to make the suppression more precse. My concern is that if the schema later becomes _totally unused_, we won't get a lint error.

**Comment by jakubmazanec** [comment]
> Either we have millions of users suffering in silence

You have multiple users complaining here. How many more do you need? 🙄

Anyway, the flag to make typeof usage valid is very simple change that you could have done instead of writing these long comments (I will probably have to do it myself, since you maintainers are so unwilling).

Also, the flag should have been added in the [original PR](https://github.com/typescript-eslint/typescript-eslint/pull/9330/, (correct me if that's wrong PR). Why? Without it it's a breaking change and should be marked as such ([it wasn't](https://github.com/typescript-eslint/typescript-eslint/blob/b17c7f25af620c22ab9a27c270c964a6fd29de7c/CHANGELOG.md?plain=1#L656), though luckily it was released among other breaking changes), because now I have linting errors in places I didn't have.

**Comment by kirkwaiblinger** [comment]
@jakubmazanec 

> You have multiple users complaining here. How many more do you need? 🙄

Believe me, we get more when we really screw things up big 😀. But this is what our "evaluating community engagement" is meant to capture, which, as indicated, I at least am open to once we better communicate recommendations around the existing behavior.

> Anyway, the flag to make typeof usage valid is very simple change that you could have done instead of writing these long comments 

This is entirely true. However, there is more to it than the work to implement a change. There's
- ongoing technical maintenance (most likely not too large in this case)
- intrinsic user burden for every option they need to learn about in order to get linting working. 
- a communication responsibility. We're aware that typescript-eslint has significant usage and that people refer to our docs to resolve differences of opinion with their teammates. Creating options comes with an implicit endorsement of their usage, so we need to be judicious about when options are created and how to appropriately document to users whether the things they report are correct 

An example of an option that should probably not have been created and therefore has created burden to maintain, difficulty in both issues and PRs for other, legitimate, options, yucky bloat on the rule's documentation page, and now, burden to deprecate as well: https://github.com/typescript-eslint/typescript-eslint/issues/9433. 

> Also, the flag should have been added in the [original PR](#9330, (correct me if that's wrong PR). Why? Without it it's a breaking change and should be marked as such ([it wasn't](https://github.com/typescript-eslint/typescript-eslint/blob/b17c7f25af620c22ab9a27c270c964a6fd29de7c/CHANGELOG.md?plain=1#L656), though luckily it was released among other breaking changes), because now I have linting errors in places I didn't have.

It's certainly not luck that it targeted the v8 branch, since it was a known breaking change/fix. But ACK that it doesn't appear to have been conspicuously documented as a breaking change (the authoritative source of truth for which is https://typescript-eslint.io/blog/announcing-typescript-eslint-v8 - the changelogs in the repo are autogenerated and not necessarily curated.) 

***

I promise you, we're not trying to be difficult, though I understand (as someone who is also opinionated about linting!) that it may feel that way. We are volunteers ourselves doing this on our personal time, and we have to balance user demand with our limited bandwidth, future maintenance effort, ease of communication, risk of bugs, etc. etc. We use a mix of the maintenance team's instincts and the input from users to inform decisions. And the current decision is document first, then consider code changes. Please work with us in that endeavor.

**Comment by jakubmazanec** [comment]
> decision is document first

Why wasn't the documentation updated in the original PR? And please don't start again about open-source, volunteering, time constraints, etc. - I contributed to some projects so I know how much time it takes. But if resources are such a problem, why did you implement this feature in a first place? My personal strategy usually is: do a proper job, or don't do it at all.

Also, if you accept my PR to add the option to turn this feature off, the behavior will be then documented simply because the option will exists (or at least I hope your documentation is, at least partially, automatically generated - if not, maybe that's where you should spend your time).

> we're not trying to be difficult

But you are. All the whining about your efforts (truly, we know, we all know that it's a hard job with ungrateful users - but maybe then don't maintain one of the most used packages). And if I actually try to help, I'm rejected. Well, maybe that's the reason you're so overwhelmed by maintenance.

**Comment by Josh-Cena** [comment]
Putting it down bluntly: a use case doesn't matter if only 0.1% of our user base requests it. A user case doesn't matter if many users want it but the team actively refuses to maintain it for any reason (at the cost of losing some users; we understand the tradeoff). Any contribution we take must be first be agreed on by the team because we are community-driven but not in an anarchy state.

> Why wasn't the documentation updated in the original PR?

We don't document every behavior of a rule; we document the _general intention_ of a rule, and behavior changes can come that are in the same spirit. For example, this rule's intention is to eliminate unused runtime code paths, so we report any dead code at runtime, including values used at compile-time. It's perfectly within many users' expectations to report it, and we even provide a specialized error message, which is also considered a form of documentation.

**Comment by jakubmazanec** [comment]
> agreed on by the team

Oh, I understand now, that's why so many of you commented on this thread, to have your little power trips :D Sorry if I don't care about programming opinions from people who wasted hours of my time and who merge PRs without documentation and then whine about how nobody helps them. If you want PR with docs, do it yourselves - I can't use documentation, what I can actually use is a fix for the feature I didn't ask for in a first place. You should be grateful instead of being smug wannabe know-it-alls.

> which is also considered a form of documentation.

That's what I said: if you merge my PR (it's 10 lines of code and 1 `if`, so please don't talk about maintenance burden, it's ridiculous), it will then be documented. If you don't, I will know that you actually are not community-driven and don't want any help.

**Comment by kirkwaiblinger** [comment]
This thread is getting heated and going in circles. Locking it to cool it down. Thanks all for the input :pray:. Will post a "next steps" comment shortly after this one.

***

Regarding @jakubmazanec specifically:


We are happy to engage in conversation that can get a bit heated at times, which can happen when people (both ourselves and community members) are passionate about their ideas. This is completely understandable, and we try to make an effort to offer our side of the story in an effort towards mutual understanding of the decision-making process, as we have done here. 

We are not happy to engage in conversation that becomes disrespectful. Name-calling, passive-aggressiveness, and other forms of personal hostility are not acceptable. It's productive to point out if you believe we've made a mistake; it's not productive to be passive-aggressive ([1](https://github.com/typescript-eslint/typescript-eslint/issues/10266#issuecomment-2578318702)) or demeaning ([2](https://github.com/typescript-eslint/typescript-eslint/issues/10266#issuecomment-2579882409)). Furthermore, on a strategic note, rudeness tends to have the effect of entrenching people in their positions rather than winning them over. This does a disservice to the other community members in this thread who have engaged in thoughtful and respectful discourse. 

Attempting to pass unsanctioned code changes under the radar is not acceptable either ([3](https://github.com/typescript-eslint/typescript-eslint/pull/10634)).

@jakubmazanez if you violate our [code of conduct](https://github.com/typescript-eslint/typescript-eslint/blob/290211b5e19cfed5dfc55baf5a18a3c8b608f8a3/CODE_OF_CONDUCT.md) again, you will be blocked from posting on this repository.

**Comment by kirkwaiblinger** [comment]
Summarizing the situation so far: in `@typescript-eslint/eslint-plugin`'s v8 release, https://github.com/typescript-eslint/typescript-eslint/pull/9330 changed `@typescript-eslint/no-unused-vars` rule to also report variables that are only used in type-positions. It previously permitted type-only uses. Ignoring those type-only uses was a longstanding feature request on the rule: https://github.com/typescript-eslint/typescript-eslint/issues/3325. 

For example, this is now reported in the rule:

```ts
import { z } from "zod";

const mySchema = z.object({ value: z.number() });
//    ~~~~~~~~
// Unused runtime variable (only used in the typeof)

export type Shape = z.infer<typeof mySchema>;
```

This is a good change _most_ of the time. _Most_ users would want this to catch unused runtime code. In that example, the usage of Zod is unnecessary and misleading. It adds bloat to the application conceptually and at runtime. It would be much less code + overhead to rewrite as e.g.:

```ts
export interface Shape {
  value: number;
}
```

The downside of this change is that some users actually want an option to let variables be only used by types. This issue has mentioned points for:

* Niche branding patterns, which work outside the standard bounds of the type system.
* Consistency when only some values are used: preferring to keep all the same, at the cost of extra syntax + runtime
* Consistency with file `import`s: e.g. when virtually all files import a helper such as `lodash`
* Types testing: for APIs that don't take in a not-used-internally runtime value

As seen in this issue, _most_, though not all, users who want the option actually can refactor to not need it. We don't believe these points are enough to justify the option being created. We instead think it should be sufficient to add docs to the rule explaining why not to let type-only values be unused.

Our next steps will be:
1. Add documentation to the rule. We'll assign this to a team member.
2. Then, if someone still wants the option, they may file a new issue that we can mark as `evaluating community engagement`.
